pensieve/config.py

```diff
@@ -0,0 +1,160 @@
+    select_expression: str
+    data_source: DataSourceReference
+
+    def resolve(self, spec: "ConfigurationSpec") -> mozanalysis.metrics.Metric:
+        return mozanalysis.metrics.Metric(
+            name=self.name,
+            data_source=self.data_source.resolve(spec),
+            select_expr=self.select_expression,
+        )
+
+
+@attr.s(auto_attribs=True)
+class MetricsConfiguration:
+    daily: List[mozanalysis.metrics.Metric]
+    weekly: List[mozanalysis.metrics.Metric]
+    overall: List[mozanalysis.metrics.Metric]
+
+
+@attr.s(auto_attribs=True)
+class MetricsSpec:
+    daily: List[MetricReference]
+    weekly: List[MetricReference]
+    overall: List[MetricReference]
+
+    definitions: Dict[str, MetricDefinition]
+
+    @classmethod
+    def from_dict(cls, d: dict) -> "MetricsSpec":
+        params = {}
+        for k in ("daily", "weekly", "overall"):
+            params[k] = d.get(k, [])
+            del d[k]
+        params["definitions"] = {k: MetricDefinition(name=k, **v) for k, v in d.items()}
+        return cls(**params)
+
+    def resolve(self, spec: "ConfigurationSpec") -> MetricsConfiguration:
+        raise NotImplementedError
+
+
+_converter.register_structure_hook(MetricsSpec, lambda obj, _type: MetricsSpec.from_dict(obj))
+
+
+@attr.s(auto_attribs=True)
+class DataSourceDefinition:
+    name: str  # implicit in configuration
+    from_expression: str
+    experiments_column_type: Optional[str]
+    client_id_column: Optional[str]
+    submission_date_column: Optional[str]
+
+    def resolve(self, spec: "ConfigurationSpec") -> mozanalysis.metrics.DataSource:
+        params = {"name": self.name, "from_expr": self.from_expression}
+        # Allow mozanalysis to infer defaults for these values:
+        for k in ("experiments_column_type", "client_id_column", "submission_date_column"):
+            v = getattr(self, k)
+            if v:
+                params[k] = v
+        return mozanalysis.metrics.DataSource(**params)
+
+
+@attr.s(auto_attribs=True)
+class DataSourcesSpec:
+    definitions: Dict[str, DataSourceDefinition]
+
+    @classmethod
+    def from_dict(cls, d: Optional[dict]) -> "DataSourcesSpec":
+        definitions = {k: DataSourceDefinition(name=k, **v) for k, v in d.items()} if d else {}
+        return cls(definitions)
+
+
+_converter.register_structure_hook(
+    DataSourcesSpec, lambda obj, _type: DataSourcesSpec.from_dict(obj)
+)
+
+
+@attr.s(auto_attribs=True)
+class AnalysisConfiguration:
+    experiment: pensieve.experimenter.Experiment
+    metrics: MetricsConfiguration
+
+
+@attr.s(auto_attribs=True)
+class ConfigurationSpec:
+    experiment: ExperimentSpec
+    metrics: MetricsSpec
+    data_sources: DataSourcesSpec
+
+    @classmethod
+    def from_dict(cls, d: dict) -> "ConfigurationSpec":
+        return _converter.structure(d, cls)
+
+    def resolve(self, experiment: pensieve.experimenter.Experiment) -> AnalysisConfiguration:
+        experiment = self.experiment.sync(experiment)
```

Comment:
Not sure about the meaning of `sync` here

Reply:
Oh, I see. It's to generate the config from a defined experiment?

Reply:
Yeah -- I thiiiiiiiiiiiiink we want to let configuration override certain properties (mostly dates) of an experiment that we learned from Experimenter.

I think we we also want context from Experimenter to interpret a config spec (mostly the platform we're targeting, so we know where to look for metrics).

So I think we need to have something that takes the info we learned from Experimenter and the assertions made by the config file and merges them into a revised Experiment object.

Reply:
I wonder if this it would make more sense the other way around: a `pensieve.Experiment` instance is created based on the info from Experimenter and then the config is applied to that instance. There'd be some `pensieve.Experiment.apply_config()` method that would update the experiment's metrics and datasources.

Reply:
Thinking about this a little more. This might be the cleaner way to do it.

Reply:
I can see either approach—I thought it might make sense to keep all of the config logic in this module but I don't have a real strong opinion.

Reply:
I kind of think a good answer is to follow the same pattern we're following for everything else, and introduce an ExperimentConfiguration class that merges the `experimenter.Experiment` and the `ExperimentSpec` -- it's confusing if `experimenter.Experiment` doesn't actually represent the information in Experimenter, and using that as our canonical experiment representation could couple us too tightly to Experimenter's API.

Reply:
> it's confusing if `experimenter.Experiment` doesn't actually represent the information in Experimenter

Agreed with this. `ExperimentConfiguration` sound like a good approach.